Replace gulp build with wp-scripts.

peterwilsoncc opened the pull request:
<!--
Filling out this template is required.  Any PR that does not include enough information to be reviewed may be closed at a maintainers' discretion.  All new code requires documentation and tests to ensure against regressions.
-->

### Description of the Change
<!--
We must be able to understand the design of your change from this description.  The maintainer reviewing this PR may not have worked with this code recently, so please provide as much detail as possible.

Where possible, please also include:
- verification steps to ensure your change has the desired effects and has not introduced any regressions
- any benefits that will be realized
- any alternative implementations or possible drawbacks that you considered
- screenshots or screencasts
-->

Replace gulp and various config files with `@wordpress/scripts`.

I think this is ready for review, it should probably get double sign-off as it is quite large. I recommend reviewing this with [whitespace suppressed](https://github.com/10up/distributor/pull/916/files?w=1).

I have put together a second webpack config for the `npm run release` command as I think it is needed to copy files in to the `release` subdirectory for committing to the `stable` branch. As noted inline, I would love an alternative suggestion as it doesn't seem ideal.

<!-- Enter any applicable Issue number(s) here that will be closed/resolved by this PR. -->
Closes #892.

### How to test the Change
<!-- Please provide steps on how to test or validate that the change in this PR works as described. -->

* Ensure production file names remain the same
* Test the various builds/npm scripts have effectively the same result (ignoring whitespace, variable munging differences)

### Changelog Entry
<!--
Please include a summary for this PR, noting whether this is something being Added / Changed / Deprecated / Removed / Fixed / or Security related.  You can replace the sample entries after this comment block with the single changelog entry line for this PR. -->
> Changed - Build process now uses `@wordpress/scripts` in place of Gulp.


### Credits
<!-- Please list any and all contributors on this PR so that they can be added to this projects CREDITS.md file. -->
Props @peterwilsoncc, @dinhtungdu 


### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you are unsure about any of these, please ask for clarification.  We are here to help! -->
- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [x] All new and existing tests pass.

peterwilsoncc committed:
Add wp-scripts as dev dependency.

peterwilsoncc committed:
Replace Gulp build process with wp-scripts.

This removes all of the previous build tools and introduces the `@wordpress/scripts` as the build tool. This is intended to increase external contributions by using a tools known within the wider WordPress community.

As a WIP, the build is expected to be broken after this commit.

peterwilsoncc committed:
Update assets as required following switch to wp-scripts.

peterwilsoncc committed:
Customize CSS Filename.

peterwilsoncc committed:
Futher update build config.

peterwilsoncc committed:
Include `.min` suffix in built files to match legacy names.

peterwilsoncc committed:
Coding standards fixes for wp-scripts.

peterwilsoncc committed:
Configure stylelint for WP Scripts.

peterwilsoncc committed:
CS Fixes: Match wp-scripts stylelint rules.

peterwilsoncc committed:
Add package.json linting.

peterwilsoncc committed:
CS: package.json.

peterwilsoncc committed:
Remove markdown linting: unsuitable for our requirements.

peterwilsoncc committed:
Global linting command.

peterwilsoncc committed:
Update JS linting to use wp-scripts.

peterwilsoncc committed:
Bump node version.

peterwilsoncc committed:
Re-add packages for building docs.

peterwilsoncc committed:
Custom release script.

peterwilsoncc committed:
Install node-wp-i18n & reinstate makepot command.

peterwilsoncc committed:
Modify release script.

peterwilsoncc commented on webpack.config.release.js:
```diff
@@ -0,0 +1,35 @@
+const CopyPlugin = require( 'copy-webpack-plugin' );
```
I'm using this in place of the gulp copy script to copy all the files over during the `release` step.

I don't really like it or think it's a proper use of webpack so suggestions on how to do this without this would be much appreciated.

peterwilsoncc committed:
Annotate ES Lints

peterwilsoncc committed:
Remove gulp config files.

peterwilsoncc committed:
Update new file name for release.

peterwilsoncc committed:
Locally define functions from document.

dinhtungdu commented on package.json:
```diff
@@ -1,68 +1,68 @@
+		"packages-update": "wp-scripts packages-update",
+		"plugin-zip": "wp-scripts plugin-zip",
+		"start": "wp-scripts start",
+		"test:e2e": "wp-scripts test-e2e",
+		"test:unit": "wp-scripts test-unit-js",
```
I think we should remove this because we don't have any e2e or unit tests.

peterwilsoncc replied:
Now the Cypress tests have been merged, I'll change the command to run those once I update against `develop`.

peterwilsoncc committed:
Update workflows using PHP or WP versions.

dinhtungdu reviewed (changes requested):
I notice some issues testing this PR:
- The build misses some files, for example, `dist/css/admin-pull-table.min.css`.
- Some JS functions are broken, for example, the Pull as draft checkbox doesn't work, see [screenshot](https://user-images.githubusercontent.com/5423135/182517096-3943ecc8-6eb4-4aff-87a4-1b278d53081a.png).
- The post editor is broken, see [screenshot](https://user-images.githubusercontent.com/5423135/182517283-c7b4439c-4a60-4c2a-90c1-7a96004898f9.png). It works fine on `develop`.
- External connection JS doesn't work too, see [screenshot](https://user-images.githubusercontent.com/5423135/182517582-b98d0468-9edd-4c4a-8e98-126d873cc3ac.png).

I will continue testing and report in follow-up comments.

dinhtungdu commented on package.json:
```diff
@@ -1,68 +1,68 @@
+		"lint:pkg-json": "wp-scripts lint-pkg-json",
+		"makepot": "wpi18n makepot && echo '.pot file updated'",
+		"packages-update": "wp-scripts packages-update",
+		"plugin-zip": "wp-scripts plugin-zip",
```
Do you think we can leverage this to create the release file?

peterwilsoncc replied:
We may be able to do so on `stable` but I will need advice on how to test it.

peterwilsoncc committed:
Merge branch 'develop' into fix/892-update-build-scripts

# Conflicts:
#	.github/workflows/lint.yml
#	package-lock.json
#	package.json

peterwilsoncc committed:
Remove wp-scripts default tests scripts.

Distributor now uses Cypress for e2e tests so the wp-scripts are unused. Existing workflows call phpunit directly so there is no need for the unit test script either.

peterwilsoncc committed:
Ignore built docs directory from linting.

peterwilsoncc committed:
Only format CSS and JS.

peterwilsoncc committed:
Auto format tests for WP JS coding standards.

peterwilsoncc committed:
Prevent Jest eslint rules triggering false positives.

peterwilsoncc committed:
Ignore false positives in script.

peterwilsoncc committed:
Document return type for cypress config.

peterwilsoncc committed:
Plugin needs to be in a JS file.

peterwilsoncc committed:
Import document as required.

peterwilsoncc committed:
Use generated file to determine dependencies.

peterwilsoncc committed:
Do not combine assets’ PHP files.

peterwilsoncc committed:
Force name of admin-pull-table CSS file.

peterwilsoncc committed:
Deal with document oddities.

peterwilsoncc committed:
Merge branch 'develop' into fix/892-update-build-scripts

# Conflicts:
#	package-lock.json
#	package.json

peterwilsoncc commented:
I think I've addressed the various issues @dinhtungdu picked up:

* fc21fd1b799e5f353f7cdf96409873e64b6ff656 fixes the block editor -- the svg needs to be a react component now 
* 104ba2322578d4ddb686cefd411db667af3f3437 fixes the name of the CSS file restoring `admin-pull-table.min.css` 
* 4cd3fc8bd40b96ad68771bba58fc7d2dbb933fa8 fixes the external connections screen

WP Scripts doesn't seem to produce code compatible with older versions of WP so I've pulled in 04a15803c6bc6e775025a39122579dbcf441321e from the requirements bump PR as both the version bump and this issue are on the same milestone.

dinhtungdu commented:
@peterwilsoncc Thanks for the updates! I can confirm all issues I noted are addressed.

I did another round of testing and found some additional issues, please check the screenshots below for more detail.

<details>
<summary>Screenshots</summary>


<img width="1368" alt="Screen Shot 2022-08-06 at 00 04 43" src="https://user-images.githubusercontent.com/5423135/183126784-27a93936-67cc-4df9-8117-4a4202347e29.png">
<img width="1368" alt="Screen Shot 2022-08-06 at 00 03 20" src="https://user-images.githubusercontent.com/5423135/183126806-566b7ec6-2863-48ce-9575-c29d51b31ba3.png">

</details>

peterwilsoncc committed:
Fix reversed media query.

peterwilsoncc commented:
Distributor panel restored in 0b53814306ae663dca9d50e6b30ad5484d6eb28c -- I'd reversed a media query.

For the `checked` attribute showing, I think that's expected due to the differences between properties and attributes.

The property shows the current state, the attribute the state at page load. If you run `jQuery('#dt-as-draft').prop( 'checked' )` in the browser console, you should get an accurate result when unchecked.

peterwilsoncc committed:
Delete correct directory when rebuilding docs.

peterwilsoncc committed:
Switch from deprecated `complete` to `always`.

peterwilsoncc commented:
@dinhtungdu I think this is ready for another review. I recommend [suppressing white space](https://github.com/10up/distributor/pull/916/files?w=1).

For now I have kept all the built asset file names the same which means the webpack config is a little odd. I've done this for back-compat but if it's decided it isn't needed then that can be done in a smaller follow up commit.

dinhtungdu reviewed (approved):
Awesome! 🎉
Thank you so much for this hard work! This PR tested well on my end.
